rs/ic_os/config/src/lib.rs

```diff
@@ -31,8 +74,170 @@ fn parse_config_line(line: &str) -> Option<(String, String)> {
+        .parse::<Ipv6Addr>()
+        .context("Invalid IPv6 gateway address")?;
+
+    let ipv4_address = config_map
+        .get("ipv4_address")
+        .map(|address| {
+            address
+                .parse::<Ipv4Addr>()
+                .context(format!("Invalid IPv4 address: {}", address))
+        })
+        .transpose()?;
+
+    let ipv4_gateway = config_map
+        .get("ipv4_gateway")
+        .map(|address| {
+            address
+                .parse::<Ipv4Addr>()
+                .context(format!("Invalid IPv4 gateway: {}", address))
+        })
+        .transpose()?;
+
+    let ipv4_prefix_length = config_map
+        .get("ipv4_prefix_length")
+        .map(|prefix| {
+            let prefix = prefix
+                .parse::<u8>()
+                .context(format!("Invalid IPv4 prefix length: {}", prefix))?;
+            if prefix > 32 {
+                bail!(
+                    "IPv4 prefix length must be between 0 and 32, got {}",
+                    prefix
+                );
+            }
+            Ok(prefix)
+        })
+        .transpose()?;
+
+    let domain = config_map.get("domain").cloned();
+
+    let networking = crate::types::NetworkSettings {
+        ipv6_prefix,
+        ipv6_address,
+        ipv6_gateway,
+        ipv4_address,
+        ipv4_gateway,
+        ipv4_prefix_length,
+        domain,
+    };
+
+    let verbose = config_map
+        .get("verbose")
+        .map(|s| s.eq_ignore_ascii_case("true"))
+        .unwrap_or(false);
+
+    Ok((networking, verbose))
+}
+
+fn default_deployment_values() -> (u32, String, Vec<Url>, String, String) {
+    (
+        490,
+        "kvm".to_string(),
+        vec![
+            Url::parse("https://icp-api.io").unwrap(),
+            Url::parse("https://icp0.io").unwrap(),
+            Url::parse("https://ic0.app").unwrap(),
+        ],
+        "mainnet".to_string(),
+        [
+            "elasticsearch-node-0.mercury.dfinity.systems:443",
+            "elasticsearch-node-1.mercury.dfinity.systems:443",
+            "elasticsearch-node-2.mercury.dfinity.systems:443",
+            "elasticsearch-node-3.mercury.dfinity.systems:443",
+        ]
+        .join(" "),
+    )
+}
+
+pub fn serialize_and_write_config<T: Serialize>(path: &Path, config: &T) -> Result<()> {
+    let serialized_config =
+        serde_json::to_string_pretty(config).expect("Failed to serialize configuration");
+    write_to_file(path, &serialized_config)
+}
+
+fn write_to_file(path: &Path, content: &str) -> Result<()> {
+    ensure_directory_exists(path)?;
+    let mut file = File::create(path)?;
+    file.write_all(content.as_bytes())?;
+    Ok(())
+}
+
+fn ensure_directory_exists(path: &Path) -> Result<()> {
+    if let Some(parent) = path.parent() {
+        if !parent.exists() {
+            create_dir_all(parent)?;
+        }
+    }
+    Ok(())
+}
+
+pub fn deserialize_config<T: for<'de> Deserialize<'de>>(file_path: &str) -> Result<T> {
```

Comment:
Is there a reason to have a custom impl instead of `serde_json::from_reader(File::open(file_path)?)`?

Reply:
Ah, much simpler! Updated!